File: app/src/main/java/com/pgshare/studentroomsharingapp/SignUp.java
package com.pgshare.studentroomsharingapp;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.Objects;

public class SignUp extends AppCompatActivity {

    private static final String TAG = "SignUpActivity";

    private TextInputLayout emailLayout, passwordLayout, confirmPasswordLayout;
    private EditText editTextEmail, passwordEditText, editTextConfirmPassword;
    private Button buttonNext;
    private ProgressBar progressBar;
    private DatabaseReference userRef;
    private FirebaseDatabase database;
    private FirebaseAuth auth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign_up);

        // Initialize Firebase
        database = FirebaseDatabase.getInstance();
        auth = FirebaseAuth.getInstance();


        // Find views
        emailLayout = findViewById(R.id.EmailLayout);
        passwordLayout = findViewById(R.id.PasswordLayout);
        confirmPasswordLayout = findViewById(R.id.ConfirmPasswordLayout);

        editTextEmail = findViewById(R.id.editTextEmail);
        passwordEditText = findViewById(R.id.passwordEditText);
        editTextConfirmPassword = findViewById(R.id.editTextConfirmPassword);

        buttonNext = findViewById(R.id.buttonNext);
        progressBar = findViewById(R.id.SignUpProgressBar);

        // Button click listener
        buttonNext.setOnClickListener(v -> onRegisterBtnClick());
    }

    private void onRegisterBtnClick() {
        String email = editTextEmail.getText().toString().trim();
        String password = passwordEditText.getText().toString().trim();
        String confirmPassword = editTextConfirmPassword.getText().toString().trim();

        // Input validation
        if (isValidInput(email, password, confirmPassword)) {
            // Show progress bar
            progressBar.setVisibility(View.VISIBLE);

            auth.createUserWithEmailAndPassword(email, password)
                    .addOnCompleteListener(this, task -> {
                        if (task.isSuccessful()) {
                            // User registration success
                            String userId = auth.getCurrentUser().getUid();


                            // Now, add user data to the Realtime Database
                            userRef = database.getReference("Users").child(userId);

                            // Replace "users" with the desired node name
                            userRef.child("email").setValue(email);


                            // You can add more data if needed, such as name, etc.
                            // database.getReference("users").child(userId).child("name").setValue(userName);

                            // Hide progress bar
                            progressBar.setVisibility(View.GONE);
                            // Navigate to the next screen
                            Intent intent = new Intent(SignUp.this, RegisterUserDetails.class);
                            startActivity(intent);

                        } else {
                            // User registration failed
                            // Handle the failure, display an error message, etc.
                            // You can check task.getException().getMessage() for the error message.
                            Objects.requireNonNull(task.getException()).getMessage();
                            // Hide progress bar
                            progressBar.setVisibility(View.GONE);
                        }
                    });

        }
    }

    private boolean isValidInput(String email, String password, String confirmPassword) {
        boolean valid = true;

        // Check if email is valid
        if (TextUtils.isEmpty(email) || !android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            emailLayout.setError("Invalid email address");
            valid = false;
        } else {
            emailLayout.setError(null);
        }

        // Check if password is empty or meets minimum length
        if (TextUtils.isEmpty(password) || password.length() < 6) {
            passwordLayout.setError("Password must be at least 6 characters");
            valid = false;
        } else {
            passwordLayout.setError(null);
        }

        // Check if passwords match
        if (!password.equals(confirmPassword)) {
            confirmPasswordLayout.setError("Passwords do not match");
            valid = false;
        } else {
            confirmPasswordLayout.setError(null);
        }

        return valid;
    }
}

File: app/src/main/java/com/pgshare/studentroomsharingapp/Authentication/SignUp.kt
package com.pgshare.studentroomsharingapp.Authentication

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import androidx.core.util.PatternsCompat
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.pgshare.studentroomsharingapp.R
import com.pgshare.studentroomsharingapp.databinding.ActivitySignUpBinding
import com.pgshare.studentroomsharingapp.viewmodel.SignUpEvent
import com.pgshare.studentroomsharingapp.viewmodel.SignUpViewModel
import kotlinx.coroutines.launch

class SignUp : AppCompatActivity() {

    private lateinit var binding: ActivitySignUpBinding
    private lateinit var viewModel: SignUpViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportActionBar?.hide()

        binding = ActivitySignUpBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = SignUpViewModel()

        binding.btnBack.setOnClickListener { finish() }
        binding.btnLoginLink.setOnClickListener { finish() }
        binding.btnCreateAccount.setOnClickListener { onRegisterBtnClick() }

        observeEvents()
    }

    private fun observeEvents() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.events.collect { event ->
                    setLoading(false)
                    when (event) {
                        is SignUpEvent.Success -> {
                            val intent = Intent(this@SignUp, Login::class.java)
                            intent.putExtra("email", binding.tilEmail.editText?.text.toString().trim())
                            startActivity(intent)
                            finish()
                        }
                        is SignUpEvent.Error -> {
                            Toast.makeText(this@SignUp, event.message, Toast.LENGTH_LONG).show()
                        }
                    }
                }
            }
        }
    }

    private fun onRegisterBtnClick() {
        val name = binding.tilName.editText?.text.toString().trim()
        val email = binding.tilEmail.editText?.text.toString().trim()
        val password = binding.tilPassword.editText?.text.toString().trim()
        val confirmPassword = binding.tilConfirmPassword.editText?.text.toString().trim()

        val role = when (binding.chipGroupRole.checkedChipId) {
            binding.chipStudent.id -> "student"
            binding.chipProfessional.id -> "professional"
            binding.chipOwner.id -> "owner"
            else -> ""
        }

        if (isValidInput(name, email, password, confirmPassword, role)) {
            setLoading(true)
            viewModel.signUp(name, email, password, role)
        }
    }

    private fun setLoading(isLoading: Boolean) {
        binding.btnCreateAccount.isEnabled = !isLoading
        binding.progressBar.visibility = if (isLoading) View.VISIBLE else View.GONE
        binding.btnCreateAccount.text = if (isLoading) "" else getString(R.string.create_account)
    }

    private fun isValidInput(name: String, email: String, password: String, confirmPassword: String, role: String): Boolean {
        var valid = true

        if (TextUtils.isEmpty(name)) {
            binding.tilName.error = "Full Name is required"
            valid = false
        } else {
            binding.tilName.error = null
        }

        if (TextUtils.isEmpty(email) || !PatternsCompat.EMAIL_ADDRESS.matcher(email).matches()) {
            binding.tilEmail.error = "Invalid email address"
            valid = false
        } else {
            binding.tilEmail.error = null
        }

        if (TextUtils.isEmpty(password) || password.length < 6) {
            binding.tilPassword.error = "Password must be at least 6 characters"
            valid = false
        } else {
            binding.tilPassword.error = null
        }

        if (password != confirmPassword) {
            binding.tilConfirmPassword.error = "Passwords do not match"
            valid = false
        } else {
            binding.tilConfirmPassword.error = null
        }

        if (!binding.cbTerms.isChecked) {
            Toast.makeText(this, "Please agree to the Terms & Conditions", Toast.LENGTH_SHORT).show()
            valid = false
        }

        if (role.isEmpty()) {
            Toast.makeText(this, "Please select if you are a Student or an Owner", Toast.LENGTH_SHORT).show()
            valid = false
        }

        return valid
    }
}
